app/src/main/java/org/mozilla/fenix/browser/OpenInAppOnboardingObserver.kt

```diff
@@ -7,62 +7,99 @@ package org.mozilla.fenix.browser
 import android.content.Context
 import android.view.ViewGroup
 import androidx.annotation.VisibleForTesting
+import androidx.lifecycle.LifecycleOwner
 import androidx.navigation.NavController
-import mozilla.components.browser.session.Session
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.mapNotNull
+import mozilla.components.browser.state.selector.selectedTab
+import mozilla.components.browser.state.store.BrowserStore
 import mozilla.components.feature.app.links.AppLinksUseCases
+import mozilla.components.lib.state.ext.flowScoped
 import mozilla.components.support.ktx.kotlin.tryGetHostFromUrl
+import mozilla.components.support.ktx.kotlinx.coroutines.flow.ifAnyChanged
 import org.mozilla.fenix.R
 import org.mozilla.fenix.ext.nav
 import org.mozilla.fenix.utils.Settings
 
 /**
  * Displays an [InfoBanner] when a user visits a website that can be opened in an installed native app.
  */
+@ExperimentalCoroutinesApi
+@Suppress("LongParameterList")
 class OpenInAppOnboardingObserver(
     private val context: Context,
+    private val store: BrowserStore,
+    private val lifecycleOwner: LifecycleOwner,
     private val navController: NavController,
     private val settings: Settings,
     private val appLinksUseCases: AppLinksUseCases,
     private val container: ViewGroup
-) : Session.Observer {
+) {
+    private var scope: CoroutineScope? = null
+    private var currentUrl: String? = null
+    private var sessionDomainForDisplayedBanner: String? = null
 
-    @VisibleForTesting
-    internal var sessionDomainForDisplayedBanner: String? = null
     @VisibleForTesting
     internal var infoBanner: InfoBanner? = null
 
-    override fun onUrlChanged(session: Session, url: String) {
-        sessionDomainForDisplayedBanner?.let {
-            if (url.tryGetHostFromUrl() != it) {
-                infoBanner?.dismiss()
+    fun start() {
+        scope = store.flowScoped(lifecycleOwner) { flow ->
+            flow.mapNotNull { state ->
+                state.selectedTab
+            }
+            .ifAnyChanged {
+                tab -> arrayOf(tab.content.url, tab.content.loading)
+            }
+            .collect { tab ->
+                if (tab.content.url != currentUrl) {
+                    sessionDomainForDisplayedBanner?.let {
+                        if (tab.content.url.tryGetHostFromUrl() != it) {
+                            infoBanner?.dismiss()
+                        }
+                    }
+                    currentUrl = tab.content.url
+                } else {
+                    // Loading state has changed
+                    maybeShowOpenInAppBanner(tab.content.url, tab.content.loading)
+                }
```

Comment:
Can there be a state where both values changed (e.g. when the selected tab has changed) and would we care about this?

Reply:
Yes, but we should be fine here, as we're only displaying the banner after a successful load. So, we want to show the banner after loading = true and loading = false happened. So if we switch to another tab with a different url we don't want to show the banner right away only if it finishes loading (if it does). If we switch to another tab with the exact same url it's fine to display the banner if needed and we haven't already displayed it for the previous tab.

Reply:
Tested this again and this actually works better than before. Now when we switch tabs to a page that can not be opened in an external app, we also don't show the banner. Before we'd also show the banner for sites where it didn't make sense if we switched tabs after the banner was displayed.